execUtil.go

```diff
@@ -3,60 +3,105 @@ package main
 import (
   "os/exec"
   "os"
-  "log"
   "strconv"
+  "time"
+  "io"
+  "bytes"
 )
 
 // handle exec calls
 
+// TODO prune finished tasks when some max map size is reached
 
 // maps task_id to cmd objects
-var TaskIdCmdMap = make(map[string]*exec.Cmd)
+var TaskIdCommandMap = make(map[string]*CommandWrapper)
 
-// TODO define Command class
+// used for task_id
+var Hostname, err = os.Hostname()
 
-func GetProcessStatus(task_id string) int {
-  log.Printf("GetProcessStatus task_id=%s, pid=%d, ProcessState=%d", task_id, TaskIdCmdMap[task_id].Process.Pid, TaskIdCmdMap[task_id].ProcessState)
-  return TaskIdCmdMap[task_id].ProcessState.ExitCode()
+type CommandWrapper struct {
+	Command   *exec.Cmd // underlying command
+	Finished  bool // set upon process finish
+  StartTime time.Time
+  EndTime   time.Time
+  Output    *bytes.Buffer
+}
+
+func GetProcessStatus(task_id string) StatusResponse {
+
+  var s StatusResponse
+
+  s.Task_id = task_id
+  s.StartTime = TaskIdCommandMap[task_id].StartTime
+  s.Finished = TaskIdCommandMap[task_id].Finished
+
+  // cmd.Wait() has finished, append
+  if(s.Finished) {
+    s.EndTime = TaskIdCommandMap[task_id].EndTime
+    s.ExitCode = TaskIdCommandMap[task_id].Command.ProcessState.ExitCode()
+  }
+
+  s.Output = TaskIdCommandMap[task_id].Output.String()
+  return s
 }
 
 // start process
-func RunCommand(command string) string {
-  log.Printf("command=%s", command)
+func RunCommand(command string) StartResponse {
+
+  var s StartResponse
+
   cmd := exec.Command(command)
-  cmd.Stdout = os.Stdout
+
+  var outBuf bytes.Buffer
+  cmd.Stdout = io.MultiWriter(os.Stdout, &outBuf)
+
   err := cmd.Start()
+
   if err != nil {
-     log.Fatal(err)
+    s.Error = err.Error()
+    return s
   }
 
+  task_id := Hostname + "-" + strconv.Itoa(cmd.Process.Pid) // TODO handle if Process or Pid nil
+  s.Task_id = task_id
+
+  TaskIdCommandMap[task_id] = &CommandWrapper{cmd,false,time.Now(),time.Time{}, &outBuf}
+
+  // async subroutine
   go func() {
-      log.Printf("pre-wait ProcessState=%v", cmd.ProcessState.ExitCode())
       err = cmd.Wait()
-      log.Printf("Command finished with error: %v", err)
-
-      // TODO mark cmd as finished in subroutine
+      TaskIdCommandMap[task_id].Finished = true
+      TaskIdCommandMap[task_id].EndTime = time.Now()
+      if err != nil  {
+        // TODO handle error
+        s.Error = err.Error()
```

Comment:
I think I understand what you mean, go maps cannot perform read/write. I will implement with a sync.Map instance, if that is a valid solution. According to documentation, this is optimized for write mostly, which begs the question of purging records on /stop call. I think the logic for now will be to keep such records, which also is easier in interpretation when considering the /status calls on finished jobs (of which stopped processes are a subset).

Reply:
What I have added is mutexes for the command map and command wrapper on get/set operations. I took the exit code and made it an attribute of the command wrapper, as this would otherwise cause data races due to the accessibility of the underlying command object (could also be resolved using go-cmd/cmd probably).